Sort placements rewrite

janm012012 opened the pull request:
# Description
This PR is a rewrite of PR #1574 by transforming the optimization into steps which are inserted into the workflow of the JobProcessor. sortLocation is removed from PnpJobPlanner and [pick|align|place]location added. This locations are precalculated in the plan() step and then used for the individual optimizations via a getter and an enum for selection.

# Justification
The old implementation was not clean and did not provide the option to extend the optimization by taking the next location into account.

# Instructions for Use
There is no change compared to PR #1574 exception that the optimization is a little better now because the next location is taken into account. Therefore jobs shall process a little faster now.

This is documentation for a user, not for a developer.

# Implementation Details
1. How did you test the change? Be descriptive. Untested code will not be accepted. **using a real job in a simulation machine**
2. Did you follow the [coding style](https://github.com/openpnp/openpnp/wiki/Developers-Guide#coding-style)? **yes**
3. If you made changes in the `org.openpnp.spi` or `org.openpnp.model` packages you will need to add additional justification for these changes. Changes to these packages require extensive review and testing. **yes: I removes PnpJobPlanner from PlannedPlacement and added other locations which are precalculated and then used for the optimization**
4. Be sure to run `mvn test` before submitting the Pull Request. If the tests do not pass the Pull Request will not be accepted. **yes**

markmaker commented:
Hi @janm012012,

I'm a bit buried in work. Can't look at this right now.

_Mark

markmaker commented:
Hi @janm012012,

Sorry, sort of forgot about this one... :sweat_smile: 

Much better now with the locators instead of the ugly `switch`, and elegantly reusing the same Locator for the end location in `optimizePlacements()`. To be entirely honest: only now do I see a real justification for the rewrite. It's really much nicer now! :sunglasses: 

Cosmetics: I would suggest some renaming to make it even clearer:

- `OptimizationNozzlesStep` -> `AbstractOptimizeNozzlesStep` (I initially misunderstood this first, due to the name)
- `getHeadLocation()` -> `convertToHeadLocation()`

I trust you tested this thoroughly in simulation?

_Mark

markmaker commented:
Brilliant. Ready to be merged?

Tested in simulation? 

Btw. what are typical gains? (assuming your simulated machine/table geometry is realistic, of course)

janm012012 commented:
Hi @markmaker!

Many thanks for reviewing!

I've renamed as suggested and made some other improvements to the readability, namely solved the FIXME to shorten the length travellingsalesman.locator by wrapping it into a class. I also fixed, that optimization was not skipped correctly in case there is only one placement.

Yes, I just rechecked the code using the simulator and I'm confident it works as expected. Except for adding the endpoint, this PR only increases the readability (IMHO).

Jan

markmaker commented:
Thanks, @janm012012 !

When I asked about "gain" I meant the gain vs. the unoptimized,, i.e. the stats you are logging.

janm012012 commented:
> When I asked about "gain" I meant the gain vs. the unoptimized,, i.e. the stats you are logging.

With optimization enabled the log of my test-job in simulation reads like

`2024-04-05 23:51:09.376 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C1, N1 (502) -> C18], 0.0% gain
2024-04-05 23:51:10.507 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N1 (502) -> C42, N2 (502b) -> C47I], 0.0% gain
2024-04-05 23:51:12.175 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 5ms: [N2 (502b) -> C47I, N1 (502) -> C42], 14.4% gain
2024-04-05 23:51:13.412 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 5ms: [N2 (502b) -> C47I, N1 (502) -> C42], 0.0% gain
2024-04-05 23:51:14.460 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 5ms: [N1 (502) -> C47H, N2 (502b) -> C47G], 0.0% gain
2024-04-05 23:51:16.094 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 5ms: [N2 (502b) -> C47G, N1 (502) -> C47H], 13.9% gain
2024-04-05 23:51:17.310 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 5ms: [N2 (502b) -> C47G, N1 (502) -> C47H], 0.0% gain
2024-04-05 23:51:18.340 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 5ms: [N2 (502b) -> C47E, N1 (502) -> C47F], 0.3% gain
2024-04-05 23:51:19.981 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 5ms: [N2 (502b) -> C47E, N1 (502) -> C47F], 0.0% gain
2024-04-05 23:51:21.227 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C47E, N1 (502) -> C47F], 0.0% gain
2024-04-05 23:51:22.263 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C47C, N1 (502) -> C47D], 0.5% gain
2024-04-05 23:51:23.893 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 5ms: [N2 (502b) -> C47C, N1 (502) -> C47D], 0.0% gain
2024-04-05 23:51:25.162 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C47C, N1 (502) -> C47D], 0.0% gain
2024-04-05 23:51:26.208 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C47A, N1 (502) -> C47B], 0.7% gain
2024-04-05 23:51:27.849 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C47A, N1 (502) -> C47B], 0.0% gain
2024-04-05 23:51:29.105 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C47A, N1 (502) -> C47B], 0.0% gain
2024-04-05 23:51:30.134 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C44, N1 (502) -> C45], 1.0% gain
2024-04-05 23:51:31.773 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C44, N1 (502) -> C45], 0.0% gain
2024-04-05 23:51:33.022 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C44, N1 (502) -> C45], 0.0% gain
2024-04-05 23:51:34.075 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 5ms: [N2 (502b) -> C41, N1 (502) -> C43], 0.1% gain
2024-04-05 23:51:35.699 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 6ms: [N2 (502b) -> C41, N1 (502) -> C43], 0.0% gain
2024-04-05 23:51:36.950 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C41, N1 (502) -> C43], 0.0% gain
2024-04-05 23:51:37.969 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 3ms: [N1 (502) -> C40, N2 (502b) -> C39], 0.0% gain
2024-04-05 23:51:39.571 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C39, N1 (502) -> C40], 14.2% gain
2024-04-05 23:51:40.844 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C39, N1 (502) -> C40], 0.0% gain
2024-04-05 23:51:41.938 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 3ms: [N1 (502) -> C36, N2 (502b) -> C35], 0.0% gain
2024-04-05 23:51:43.560 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C35, N1 (502) -> C36], 11.1% gain
2024-04-05 23:51:44.836 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 5ms: [N1 (502) -> C36, N2 (502b) -> C35], 11.8% gain
2024-04-05 23:51:45.877 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C32, N1 (502) -> C33], 3.3% gain
2024-04-05 23:51:47.492 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 5ms: [N2 (502b) -> C32, N1 (502) -> C33], 0.0% gain
2024-04-05 23:51:48.744 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N1 (502) -> C33, N2 (502b) -> C32], 11.9% gain
2024-04-05 23:51:49.799 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 5ms: [N2 (502b) -> C30, N1 (502) -> C31], 0.2% gain
2024-04-05 23:51:51.436 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C30, N1 (502) -> C31], 0.0% gain
2024-04-05 23:51:52.692 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 5ms: [N2 (502b) -> C30, N1 (502) -> C31], 0.0% gain
2024-04-05 23:51:53.697 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 6ms: [N1 (502) -> C26, N2 (502b) -> C25], 0.0% gain
2024-04-05 23:51:55.315 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 6ms: [N2 (502b) -> C25, N1 (502) -> C26], 14.8% gain
2024-04-05 23:51:56.575 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C25, N1 (502) -> C26], 0.0% gain
2024-04-05 23:51:57.574 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 8ms: [N1 (502) -> C23, N2 (502b) -> C21], 0.0% gain
2024-04-05 23:51:59.217 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 6ms: [N2 (502b) -> C21, N1 (502) -> C23], 14.9% gain
2024-04-05 23:52:00.458 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C21, N1 (502) -> C23], 0.0% gain
2024-04-05 23:52:01.458 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 3ms: [N1 (502) -> C19, N2 (502b) -> C17], 0.0% gain
2024-04-05 23:52:03.088 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 6ms: [N2 (502b) -> C17, N1 (502) -> C19], 15.1% gain
2024-04-05 23:52:04.319 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C17, N1 (502) -> C19], 0.0% gain
2024-04-05 23:52:05.347 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 5ms: [N1 (502) -> C16, N2 (502b) -> C15], 0.0% gain
2024-04-05 23:52:06.966 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C15, N1 (502) -> C16], 15.1% gain
2024-04-05 23:52:08.225 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C15, N1 (502) -> C16], 0.0% gain
2024-04-05 23:52:09.283 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 3ms: [N1 (502) -> C14, N2 (502b) -> C13], 0.0% gain
2024-04-05 23:52:10.902 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C13, N1 (502) -> C14], 12.0% gain
2024-04-05 23:52:12.142 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C13, N1 (502) -> C14], 0.0% gain
2024-04-05 23:52:13.198 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C11, N1 (502) -> C12], 1.3% gain
2024-04-05 23:52:14.842 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C11, N1 (502) -> C12], 0.0% gain
2024-04-05 23:52:16.089 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C11, N1 (502) -> C12], 0.0% gain
2024-04-05 23:52:17.166 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 3ms: [N2 (502b) -> C9, N1 (502) -> C10], 1.7% gain
2024-04-05 23:52:18.825 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C9, N1 (502) -> C10], 0.0% gain
2024-04-05 23:52:20.054 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N1 (502) -> C10, N2 (502b) -> C9], 9.8% gain
2024-04-05 23:52:21.131 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 3ms: [N2 (502b) -> C7, N1 (502) -> C8], 3.6% gain
2024-04-05 23:52:22.783 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C7, N1 (502) -> C8], 0.0% gain
2024-04-05 23:52:24.013 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N1 (502) -> C8, N2 (502b) -> C7], 1.3% gain
`

I guess I'll have to implement a milometer to measure the different in more details. Unfortunately I does not look that spectacular in the logs. However, on my machine, where N2 shall always be handled first in alignment, the movement process looks much more natural with this optimization...

Jan

janm012012 commented:
Hmm... I just noticed, that the calculated gain references the previous state instead of the unoptimized state as plannedPlacements are hand over from one step to the next preserving their order. This means that the order optimized for picking is taken as reference for the alignment optimization and the one optimized for alignment as reference for the place optimization when calculating the indicated gain. This means for better logging plannedPlacements shall be reverted to their original sorting order first or the gain factor shall be omitted. 

With the original sorting order as base my test job read like this now:

`2024-04-06 07:11:00.365 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 7ms: [N2 (502b) -> C47I, N1 (502) -> C42], 14.5% gain
2024-04-06 07:11:01.574 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C47I, N1 (502) -> C42], 8.4% gain
2024-04-06 07:11:02.620 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 5ms: [N2 (502b) -> C47G, N1 (502) -> C47H], 0.0% gain
2024-04-06 07:11:04.282 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C47G, N1 (502) -> C47H], 14.5% gain
2024-04-06 07:11:05.486 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C47G, N1 (502) -> C47H], 6.5% gain
2024-04-06 07:11:06.493 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 10ms: [N2 (502b) -> C47E, N1 (502) -> C47F], 0.3% gain
2024-04-06 07:11:08.196 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C47E, N1 (502) -> C47F], 14.0% gain
2024-04-06 07:11:09.383 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C47E, N1 (502) -> C47F], 5.3% gain
2024-04-06 07:11:10.404 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C47C, N1 (502) -> C47D], 0.5% gain
2024-04-06 07:11:12.073 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C47C, N1 (502) -> C47D], 13.5% gain
2024-04-06 07:11:13.309 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C47C, N1 (502) -> C47D], 4.2% gain
2024-04-06 07:11:14.334 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C47A, N1 (502) -> C47B], 0.7% gain
2024-04-06 07:11:15.979 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C47A, N1 (502) -> C47B], 12.9% gain
2024-04-06 07:11:17.154 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C47A, N1 (502) -> C47B], 2.7% gain
2024-04-06 07:11:18.181 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 5ms: [N2 (502b) -> C44, N1 (502) -> C45], 1.0% gain
2024-04-06 07:11:19.845 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C44, N1 (502) -> C45], 15.4% gain
2024-04-06 07:11:21.016 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C44, N1 (502) -> C45], 12.9% gain
2024-04-06 07:11:22.016 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C41, N1 (502) -> C43], 0.1% gain
2024-04-06 07:11:23.696 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 12ms: [N2 (502b) -> C41, N1 (502) -> C43], 15.9% gain
2024-04-06 07:11:24.894 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C41, N1 (502) -> C43], 8.8% gain
2024-04-06 07:11:25.903 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 6ms: [N1 (502) -> C40, N2 (502b) -> C39], 0.0% gain
2024-04-06 07:11:27.544 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 5ms: [N2 (502b) -> C39, N1 (502) -> C40], 14.2% gain
2024-04-06 07:11:28.757 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C39, N1 (502) -> C40], 11.9% gain
2024-04-06 07:11:29.842 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 5ms: [N1 (502) -> C36, N2 (502b) -> C35], 0.0% gain
2024-04-06 07:11:31.491 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C35, N1 (502) -> C36], 11.2% gain
2024-04-06 07:11:32.695 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N1 (502) -> C36, N2 (502b) -> C35], 0.0% gain
2024-04-06 07:11:33.687 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C32, N1 (502) -> C33], 3.5% gain
2024-04-06 07:11:35.349 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 5ms: [N2 (502b) -> C32, N1 (502) -> C33], 14.8% gain
2024-04-06 07:11:36.655 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N1 (502) -> C33, N2 (502b) -> C32], 0.0% gain
2024-04-06 07:11:37.698 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C30, N1 (502) -> C31], 0.2% gain
2024-04-06 07:11:39.339 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C30, N1 (502) -> C31], 15.8% gain
2024-04-06 07:11:40.529 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C30, N1 (502) -> C31], 8.7% gain
2024-04-06 07:11:41.526 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N1 (502) -> C26, N2 (502b) -> C25], 0.0% gain
2024-04-06 07:11:43.200 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C25, N1 (502) -> C26], 14.9% gain
2024-04-06 07:11:44.452 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C25, N1 (502) -> C26], 5.2% gain
2024-04-06 07:11:45.444 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N1 (502) -> C23, N2 (502b) -> C21], 0.0% gain
2024-04-06 07:11:47.096 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C21, N1 (502) -> C23], 15.0% gain
2024-04-06 07:11:48.296 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C21, N1 (502) -> C23], 6.7% gain
2024-04-06 07:11:49.295 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 3ms: [N1 (502) -> C19, N2 (502b) -> C17], 0.0% gain
2024-04-06 07:11:50.953 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C17, N1 (502) -> C19], 15.2% gain
2024-04-06 07:11:52.149 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N2 (502b) -> C17, N1 (502) -> C19], 14.5% gain
2024-04-06 07:11:53.155 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N1 (502) -> C16, N2 (502b) -> C15], 0.0% gain
2024-04-06 07:11:54.845 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C15, N1 (502) -> C16], 15.1% gain
2024-04-06 07:11:56.036 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 10ms: [N2 (502b) -> C15, N1 (502) -> C16], 15.1% gain
2024-04-06 07:11:57.058 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N1 (502) -> C14, N2 (502b) -> C13], 0.0% gain
2024-04-06 07:11:58.704 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C13, N1 (502) -> C14], 12.1% gain
2024-04-06 07:11:59.893 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C13, N1 (502) -> C14], 6.9% gain
2024-04-06 07:12:00.914 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 3ms: [N2 (502b) -> C11, N1 (502) -> C12], 1.3% gain
2024-04-06 07:12:02.601 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C11, N1 (502) -> C12], 12.2% gain
2024-04-06 07:12:03.819 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 3ms: [N2 (502b) -> C11, N1 (502) -> C12], 14.9% gain
2024-04-06 07:12:04.879 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 3ms: [N2 (502b) -> C9, N1 (502) -> C10], 1.7% gain
2024-04-06 07:12:06.591 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C9, N1 (502) -> C10], 11.1% gain
2024-04-06 07:12:07.798 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 4ms: [N1 (502) -> C10, N2 (502b) -> C9], 0.0% gain
2024-04-06 07:12:08.841 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for pick completed in 4ms: [N2 (502b) -> C7, N1 (502) -> C8], 3.7% gain
2024-04-06 07:12:10.548 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for alignment completed in 4ms: [N2 (502b) -> C7, N1 (502) -> C8], 12.5% gain
2024-04-06 07:12:11.720 ReferencePnpJobProcessor$AbstractOptimizationNozzlesStep DEBUG: Optimization for place completed in 5ms: [N1 (502) -> C8, N2 (502b) -> C7], 0.0% gain
`  

It now correctly indicates, that alignment is always much better when executed in the order N2, N1.

There is one more thing that I do not understand yet: according to the gains and to the understanding of my machine, a job with optimization shall run faster compared to the original unoptimized state. However, in simulation I can't see that. Locking that the execution time and CPHs reported my test needs 77.6s optimized and 77.8s without. I'd say thats insignificant. Could it be, that the simulation driver does not correctly simulate the execution time of motion? Or do you have any other explanation for that?

However, despite this measurements I'm still convinced the optimization is advantageous, at least for me...

Jan

markmaker commented:
These savings are **not** negligible. :+1: 

Currently you are logging linear travel gains. As we discussed earlier, for most moves that do not reach top speed, due to acceleration/deceleration ramping, the move times are not linear but to the square root of the move distances, i.e. four times the move distance will only take twice the time. So your logged percentages are diminished in reality.

Then there's all the non-optimized (e.g. Z) movement. And non-move time for vision, vacuum, etc. 

Plus, as discussed in the predecessor PR, the real big gain in optimization would come from wholly reordering the placements (not just the "nozzle-full"). Then one could optimize for feeder co-location etc.

Gotta go now... :wink: 

_Mark

janm012012 commented:
> Currently you are logging linear travel gains. As we discussed earlier, for most moves that do not reach top speed, due to acceleration/deceleration ramping, the move times are not linear but to the square root of the move distances, i.e. four times the move distance will only take twice the time. So your logged percentages are diminished in reality.
> 
That's on my list for an other PR.

> Plus, as discussed in the predecessor PR, the real big gain in optimization would come from wholly reordering the placements (not just the "nozzle-full"). Then one could optimize for feeder co-location etc.
> 
Based on what is now available with this PR, I'll preparing a PR that selects the best (based on distance between pick and place locations) next placement with respect to the ones already selected. That shall be the next intermediate step towards perfectly optimized job execution.

Jan

PS: I've fixed the gain logging locally. If you think, it shall be part of the official code base now, I'll convert it into a PR.

markmaker commented:
> Based on what is now available with this PR, I'll preparing a PR that selects the best (based on distance between pick and place locations) next placement with respect to the ones already selected. That shall be the next intermediate step towards perfectly optimized job execution.

I doubt it will ever be possible to optimize realistically large and heterogeneous jobs "perfectly". You will likely have to opt for a [greedy algorithm](https://en.wikipedia.org/wiki/Greedy_algorithm) and with realistically mixed nozzle tip compatibilities, and strategic nozzle tip changing part of the equation, it will likely be far from perfect. :rofl:

>  I've fixed the gain logging locally. 

You could piggy back this small change on your subordinal motion PR.

(I'm still only sporadically online)

_Mark
